Source/DafnyCore/Resolver/PreType/PreTypeResolve.State.cs

```diff
@@ -0,0 +1,886 @@
+              }
+            } else {
+              yield return (DPreType)sub;
+            }
+          }
+        }
+      }
+    }
+
+    IEnumerable<DPreType> AllSuperBounds(PreTypeProxy proxy, ISet<PreTypeProxy> visited) {
+      Contract.Requires(proxy.PT == null);
+      if (!visited.Contains(proxy)) {
+        visited.Add(proxy);
+        foreach (var constraint in unnormalizedSubtypeConstraints) {
+          if (constraint.Sub.Normalize() == proxy) {
+            var super = constraint.Super.Normalize();
+            if (super is PreTypeProxy superProxy) {
+              foreach (var pt in AllSuperBounds(superProxy, visited)) {
+                yield return pt;
+              }
+            } else {
+              yield return (DPreType)super;
+            }
+          }
+        }
+      }
+    }
+
+
+    IEnumerable<PreType> DirectionalBounds(PreTypeProxy tproxy, int direction) {
+      foreach (var constraint in unnormalizedSubtypeConstraints) {
+        if (0 <= direction && constraint.Super.Normalize() == tproxy) {
+          yield return constraint.Sub;
+        }
+        if (direction <= 0 && constraint.Sub.Normalize() == tproxy) {
+          yield return constraint.Super;
+        }
+      }
+    }
+
+    // ---------------------------------------- Guarded constraints ----------------------------------------
+
+    private List<Func<bool>> guardedConstraints = new();
+
+    void AddGuardedConstraint(Func<bool> predicate) {
+      Contract.Requires(predicate != null);
+      guardedConstraints.Add(predicate);
+    }
+
+    bool ApplyGuardedConstraints() {
+      if (guardedConstraints.Count == 0) {
+        // common special case
+        return false;
+      }
+      var constraints = guardedConstraints;
+      guardedConstraints = new();
+      var anythingChanged = false;
+      foreach (var constraint in constraints) {
+        if (constraint()) {
+          anythingChanged = true;
+        } else {
+          guardedConstraints.Add(constraint);
+        }
+      }
+      return anythingChanged;
+    }
+
+    // ---------------------------------------- Advice ----------------------------------------
+
+    class Advice {
+      public readonly PreType PreType;
+      public readonly AdviceTarget What;
+
+      public string WhatString => What == AdviceTarget.Object ? "object?" : What.ToString().ToLower();
+
+      public Advice(PreType preType, AdviceTarget advice) {
+        PreType = preType;
+        What = advice;
+      }
+    }
+
+    enum AdviceTarget {
+      Bool, Char, Int, Real, String, Object
+    }
+
+    private List<Advice> defaultAdvice = new();
+
+    void AddDefaultAdvice(PreType preType, AdviceTarget advice) {
+      Contract.Requires(preType != null);
+      defaultAdvice.Add(new Advice(preType, advice));
+    }
+
+    bool ApplyDefaultAdvice() {
+      bool anythingChanged = false;
+      foreach (var advice in defaultAdvice) {
+        var preType = advice.PreType.Normalize();
```

Comment:
Can you move the body of the loop into a method that's inside the class `Advice`, and move that class into a separate file together with `AdviceTarget`?

Also, consider moving `AdviceTarget` inside `Advice` and renaming it to `Target`, so you reference it using `Advice.Target`